design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,632 @@
+to be set per secret instance. For example, `labels`, `annotations` and `type`
+of the secret in _Kubernetes_; `Tags` and `EncryptionKey` in _AWS_.
+
+A `StoreConfig` named `default` will be created during installation which is
+configured to write secret to a Kubernetes cluster in the Crossplane
+installation namespace. 
+
+`StoreConfig.configRef` will be an optional parameters and **in its absence**:
+
+- For cluster-scoped resources (e.g. `MR` and `XR`), it will be late initialized
+as `configRef.name=default`.
+- For namespaced resources (e.g. `XRC`), secrets published to the same
+namespace as the resource.
+
+**Examples:**
+
+Publish Connection details to a _Kubernetes_ secret (already existing case):
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    configRef:
+      name: default
+```
+
+Publish Connection details to a _Vault_ secret:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    configRef:
+      name: vault-dev
+```
+
+Publish Connection details to a _Kubernetes_ secret with some labels/annotations:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    attributes:
+      labels:
+        environment: production
+      annotations:
+        acme.example.io/secret-type: infrastructure
+    configRef:
+      name: default
+```
+
+Publish Connection details to an _AWS Secret Manager_ secret with some tags:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    attributes:
+      tags:
+        environment: production
+    configRef:
+      name: aws-secret-manager-prod
+```
+
+#### Separation of Concerns: Compositions, Composites and Claims
+
+Currently, platform operators could specify where should composite resource
+secrets land using the partial input `writeConnectionSecretsToNamespace`. We
+will similarly deprecate this field in favor of
+`publishConnectionDetailsToStore`.
+
+**Examples:**
+
+**Composition** configuring to publish to default Store which is created during
+installation time, i.e. publish to a Kubernetes secret in the namespace where
+crossplane installed.
+
+```yaml
+spec:
+  publishConnectionDetailsToStore: default
+```
+
+**Composition** configuring to publish to another namespace, e.g. 
+`infrastructure-staging`, where a `StoreConfig` named
+`store-infrastructure-staging` created with namespace parameter as
+`infrastructure-staging`.
+
+```yaml
+spec:
+  publishConnectionDetailsToStore: store-infrastructure-staging
+```
+
+**Composition** configuring to publish to vault.
+
+```yaml
+spec:
+  publishConnectionDetailsToStore: vault-production
+```
+
+**Claim** resources requires some additional logic since they are namespaced but
+need a Cluster scoped `StoreConfig`. They will always use the `default`
+`StoreConfig` and if it is configured for local Kubernetes, secret will be
+written into the **Claim** namespace instead of namespace configured in `default`
+`StoreConfig`.
```

Comment:
@negz this is the best way I could come to handle referring a cluster scoped `StoreConfig` from a namespaced `Claim` and I cannot say that I am happy with it.

Apart from having some implicit logic that differs for Kubernetes and other secret stores, the main limitation of this approach is we are losing the scoping done at the namespace level. As a concrete example, if the `default` store is configured as `Vault`, secrets in namespace `team-1` and `team-2` would conflict. 

I believe we could solve this by introducing a _scope_ concept to Store implementations. Usage of scope would differ for each store implementation, for example, for Kubernetes it will just be used as the namespace of the secret but for vault, it will be appended to the parentPath of the secret. This way, we could define a common `defaultScope` parameter for `StoreConfig` resource which would be used whenever a scope is not available (i.e. cluster scoped).

```
apiVersion: secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: default
spec:
  type: Kubernetes
  defaultScope: crossplane-system
  kubernetes:
    ...
```

```
apiVersion: secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: default
spec:
  type: Vault
  defaultScope: crossplane-system
  vault:
    parentPath: "secret/my-cloud/dev/"
    ...
```

or we can just name is a `defaultNamespace`. What do you think?

Reply:
I think your proposal sounds good.

My only question would be whether we're confident every secret store we'll support has a concept of scope? If there was any concern about that we could move the `defaultScope` into the more specific store type blocks, like:

```yaml
apiVersion: secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: default
spec:
  type: Kubernetes
  kubernetes:
    # Used for cluster scoped types. Namespaced types always use their own namespace.
    defaultNamespace: crossplane-system
```

```yaml
apiVersion: secrets.crossplane.io/v1alpha1
kind: StoreConfig
metadata:
  name: default
spec:
  type: Vault
  vault:
    parentPath: "secret/my-cloud-dev"
    # Used for cluster scoped types. Namespaced types always use their own namespace.
    defaultParentPathSuffix: crossplane-system
```

I don't feel strongly about this though - if you're confident scope will always be relevant making it applicable to all secret stores makes sense. I suppose even if a secret store didn't have a native concept of scope we could always prefix it to the secret name in that store, so name the secret something like `crossplane-system-foo`.

Reply:
Yeah that's a good point. I believe any secret store we are adding support for, needs to somehow handle secrets coming from multiple namespaces with some sort of scoping which could be just prefixing to secret name as you suggested. So, I will add a `spec.defaultScope` to `StoreConfig`.